### PR TITLE
[8.x] Add support for multiple connections to parallel testing

### PR DESCRIPTION
**Pull request by ibrasho**
The current parallel testing implementation assumes only the default connection needs to be switch to test databases.

This PR adds a new config key (`database.testing.connections`) that be an array of connection names that should be switch to the current database while testing.

### CODE REVIEW

**Comment by ibrasho** [comment]
Not sure if the config name is suitable or if it even belongs in `database`.

**Comment by mfn** [comment]
A fix for this would be very welcoming, as it's the reason I can't even try this feature.

File: src/Illuminate/Testing/Concerns/TestDatabases.php [hunk 1]
```diff
@@ -146,22 +146,24 @@ protected function whenNotUsingInMemoryDatabase($callback)
...
-            config()->set(
-                "database.connections.{$default}.url",
-                preg_replace('/^(.*)(\/[\w-]*)(\??.*)$/', "$1/{$database}$3", $url),
-            );
-        } else {
-            config()->set(
-                "database.connections.{$default}.database",
-                $database,
-            );
+        $connections = config('database.testing.connections', [config('database.default')]);
```
Comment (mfn) [inline src/Illuminate/Testing/Concerns/TestDatabases.php:149]: Something like `database.testing` does not exist?

Not sure if it's feasible to to something like https://github.com/laravel/framework/blob/a32a32f628eec0cee87d4e7d4b32891cfb45b2e4/src/Illuminate/Foundation/Testing/DatabaseTransactions.php#L16
which effectively makes use of a property `connectionsToTransact`

**Comment by taylorotwell** [comment]
Should be a property on the base test class IMO - not a configuration item.

**Comment by tomirons** [comment]
@ibrasho it doesn't appear that this sets database names for different connections. For example if my main database is `laravel` and I have another connection with the database name of `alternate`, does this overwrite the database name to be `laravel_test_1` for both connections? I would expect the alternate connection to have a database of `alternate_test_1`.

File: src/Illuminate/Testing/Concerns/TestDatabases.php [hunk 2]
```diff
@@ -146,22 +146,24 @@ protected function whenNotUsingInMemoryDatabase($callback)
...
+
+        foreach ($connections as $connection) {
```
Comment (tomirons) [inline src/Illuminate/Testing/Concerns/TestDatabases.php:151]: @ibrasho it doesn't appear that this sets database names for different connections. For example if my main database is `laravel` and I have another connection with the database name of `alternate`, does this overwrite the database name to be `laravel_test_1` for both connections? Shouldn't the alternate connection have a database name of `alternate_test_1`.

**Review by tomirons — CHANGES_REQUESTED** [review]

**Comment by nunomaduro** [comment]
Thanks for your pull request to Laravel!

Unfortunately, we are going to delay merging this code for now. To preserve our ability to adequately maintain the framework, we need to be very careful regarding the amount of code we include.

If eventually more people have the need for this feature, we will re-consider it.

**Comment by taylorotwell** [comment]
I'm not opposed to having this feature, but it would need to be a property on the base test class.

**Comment by nunomaduro** [comment]
@ibrasho Can you apply Taylor's suggestion? I will analyze the pull request after that.

File: src/Illuminate/Testing/Concerns/TestDatabases.php (hunk 1 above)
Comment (ibrasho) [reply src/Illuminate/Testing/Concerns/TestDatabases.php:149]: I was under the impression that we don't have access to the test case instance. I'm updating this PR with a cleaner implementation.

File: src/Illuminate/Testing/Concerns/TestDatabases.php (hunk 2 above)
Comment (ibrasho) [reply src/Illuminate/Testing/Concerns/TestDatabases.php:151]: I'm using Postgres schemas for the landlord and multiple tenants. I'm sure in other cases the requirements would be different.

What you are suggesting could also work, I'll try to see if I can make the logic simple and versatile at the same time.

**Comment by tomirons** [comment]
@ibrasho are you going to add the necessary logic to update the database names as I had explained? If not I'll find time to do it myself after this has been merged.

**Comment by ibrasho** [comment]
@tomirons, I'm trying to work out the logic to make that happen. @taylorotwell suggested adding a property to the base class, but what you suggested requires creating multiple databases in the `setUpProcess` callback (which is run before any test case is instantiated. I'm thinking of using a static property on the base class. Would that be fine, @taylorotwell?

The best option is to provide the user with the ability to override the default logic instead of catering to every scenario. In my case, I had to override 3 classes to be able to do this. If the framework gave the users the ability to override the default logic easily, things should be better.

I should have some time to finish this later today.

**Comment by tomirons** [comment]
@ibrasho I completely understand taking the robust approach here and I support it. I was just voicing my concern on how tight knit the current implementation was, that's all. 😃

**Comment by taylorotwell** [comment]
There seem to be some unresolved questions and concerns around this PR and I don't want to take on the maintenance burden of maintaining a bad implementation if this isn't solving everyone's problems.

**Comment by ibrasho** [comment]
@taylorotwell, I agree, and I'm trying to find an implementation that introduces the least amount of complexity. The real issue is that testing logic with multiple databases is different per project, which complicated the solution.

The simplest solution requiring the least amount of work is to provide a way to override the callbacks provided to `ParallelTesting::setUpProcess` and `ParallelTesting::setUpTestCase` in `Illuminate\Testing\Concerns\TestDatabases`. I implemented this in a [new branch](https://github.com/ibrasho-forks/framework/commit/853e2b5676e708447f981029174068e25dc4e4e3), so you see the changes proposed.

It's very minimal and pushes the complexity to the implementer (which much help since they need to reimplement most of `TestDatabases` methods, honestly).

The other alternative is what I just pushed, making `TestDatabases` more robust and providing more tools to the end-user. It adds a new static property to `TestDatabases::$connections`, an array of connections to manage. All logic in `TestDatabases` are updated to depend on that property. The end-user would need to set this in a provider for example `ParallelTestingServiceProvider::$connections = ['c1', 'c2'];`. You can [see this implementation here](https://github.com/ibrasho-forks/framework/commit/60bb47f1300811e0dbcbc140a42c562a01037155).

Let me know if you think either is an acceptable solution for you.

**Comment by tomirons** [comment]
@ibrasho I really like the second implementation a lot. It definitely gives the most flexibility. 👍 

Here are just a few things I saw when looking through it.

1. There are quite a few `dump`'s lying around. Not sure if they were missed but I thought I'd let you know.
    - [one](https://github.com/ibrasho-forks/framework/commit/60bb47f1300811e0dbcbc140a42c562a01037155?branch=60bb47f1300811e0dbcbc140a42c562a01037155&diff=split#diff-e14de3b8b89927a4173ce411b40520d70feb289b071aa0b7d6cdfb043f3fc696R57)
    - [two](https://github.com/ibrasho-forks/framework/commit/60bb47f1300811e0dbcbc140a42c562a01037155?branch=60bb47f1300811e0dbcbc140a42c562a01037155&diff=split#diff-e14de3b8b89927a4173ce411b40520d70feb289b071aa0b7d6cdfb043f3fc696R136)
    - [three](https://github.com/ibrasho-forks/framework/commit/60bb47f1300811e0dbcbc140a42c562a01037155?branch=60bb47f1300811e0dbcbc140a42c562a01037155&diff=split#diff-e14de3b8b89927a4173ce411b40520d70feb289b071aa0b7d6cdfb043f3fc696R141)
2. Do you think prepending the connection name is appropriate [here](https://github.com/ibrasho-forks/framework/commit/60bb47f1300811e0dbcbc140a42c562a01037155?branch=60bb47f1300811e0dbcbc140a42c562a01037155&diff=split#diff-e14de3b8b89927a4173ce411b40520d70feb289b071aa0b7d6cdfb043f3fc696R239)?
3. I would rename the `$connection` variable to `$database`  [here](https://github.com/ibrasho-forks/framework/commit/60bb47f1300811e0dbcbc140a42c562a01037155?branch=60bb47f1300811e0dbcbc140a42c562a01037155&diff=split#diff-e14de3b8b89927a4173ce411b40520d70feb289b071aa0b7d6cdfb043f3fc696R156)

**Comment by tomirons** [comment]
@ibrasho after testing out your changes with our application that has over 4k tests and has multiple different connections. It didn't seem to work at all. 😢

**Comment by ibrasho** [comment]
@tomirons Can you clarify the issue you face? Which changes did you try? I've used something similar in one of my projects and it's working fine. I might write a blog post on how to do this in a project if this doesn't get merged.

**Comment by tomirons** [comment]
@ibrasho I applied the second implementation that you linked to [here](https://github.com/ibrasho-forks/framework/commit/60bb47f1300811e0dbcbc140a42c562a01037155).

1. Not sure where to set the _connections_ at. In the `AppServiceProvider`, elsewhere?
2. On the _schema is up to date_ [check](https://github.com/ibrasho-forks/framework/commit/60bb47f1300811e0dbcbc140a42c562a01037155#diff-e14de3b8b89927a4173ce411b40520d70feb289b071aa0b7d6cdfb043f3fc696R155) it was failing as well because it couldn't find the connection in that array.

I would love to get this working for our application. 🙏

**Comment by CraigMonva** [comment]
Have a test suite thats taking 30 minutes to run... finally getting around to using parallel testing in our app that makes use of mysql and mongo connections but the framework only handles using test databases for the default connection.

I can imagine this feature being requested by more of the community over time.

Any updates on this?

**Comment by robsontenorio** [comment]
@ibrasho  @tomirons 

Do you guys have found any solution? I have a similar issue with Postgres.

The problem with Postgres is when you use a schema name different than `public`.

- Works : php artisan test
- Works: php artisan test --parallel

```
'pgsql' => [
            'driver' => 'pgsql',
            ....
            'schema' => 'public',
        ],
```


- Works : php artisan test
- **Dont Work**: php artisan test --parallel

```
'pgsql' => [
            'driver' => 'pgsql',
            ....
            'schema' => 'myblog',
        ],
```

Paratest always tries to use `public` schema.

This is a problem, because sometimes you don't want to use `public` schema for security reasons.

**Comment by tomirons** [comment]
@robsontenorio I have not come up with a solution to this issue. The `schema` problem would be a separate issue than what is trying to be solved here.

**Comment by robsontenorio** [comment]
@tomirons Just opened an issue #39775  but just noticed there is no interest at this moment to include that feature on framework. Maybe in the future, but for now,  I am removing parallel tests on my codebase.

**Comment by Orrison** [comment]
@taylorotwell @nunomaduro I would like to implement this feature in a possibly simpler and easier to maintain way. I have a working implementation I think will work for y'all. The only issue is the storage location of the array of connection names.

After some tinkering, I am unsure of how it would be possible to set this as a property of the Testcase as it seems not possible to resolve what the Testcase class is during the boot of `ParallelTestingServiceProvider`.
Do you suggest some way of resolving the used Testcase during `ParallelTestingServiceProvider` and/or the ParallelTesting callbacks (other than `ParallelTesting::setUpTestCase`)

This would be an extremely useful and powerful feature for the codebase that I feel we are currently missing.

**Comment by Orrison** [comment]
Currently it seems impossible since `callSetUpProcessCallbacks` is called before your Testcase is created/setup.

**Pull request closed without merging**